File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/EndsWithEvaluator.java
package org.opencds.cqf.cql.engine.elm.execution;

import org.opencds.cqf.cql.engine.execution.Context;

public class EndsWithEvaluator extends org.cqframework.cql.elm.execution.EndsWith {

    public static Object endsWith(String argument, String suffix) {
        if (argument == null || suffix == null) {
            return null;
        }
        return argument.endsWith(suffix);
    }

    @Override
    protected Object internalEvaluate(Context context) {
        String argument = (String) getOperand().get(0).evaluate(context);
        String suffix = (String) getOperand().get(1).evaluate(context);

        return endsWith(argument, suffix);
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/EndsWithEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.runtime.Boolean
import org.opencds.cqf.cql.engine.runtime.String
import org.opencds.cqf.cql.engine.runtime.Value
import org.opencds.cqf.cql.engine.runtime.toCqlBoolean

object EndsWithEvaluator {
    @JvmStatic
    fun endsWith(argument: Value?, suffix: Value?): Boolean? {
        if (argument == null || suffix == null) {
            return null
        }

        if (argument is String && suffix is String) {
            return argument.endsWith(suffix).toCqlBoolean()
        }

        throw InvalidOperatorArgument(
            "EndsWith(String, String)",
            "EndsWith(${argument.typeAsString}, ${suffix.typeAsString})",
        )
    }
}
